### PR TITLE
[Core] Logprobs support in Multi-step

### PR DESCRIPTION
**Pull request by afeldman-nm**
**NOTE**: because `SamplerOutput` was refactored from `sequence.py` into `sampler.py`, this PR makes a large number of trivial import changes to all files which use `SamplerOutput` (i.e. many of the model files.) **The only files you need to review are**:
* `tests/multi-step/test_correctness.py`
* `vllm/model_executor/layers/sampler.py`
* `vllm/worker/multi_step_model_runner.py`

----

The purpose of this PR is to achieve parity between the logprobs UX for multi-step scheduling, and the logprobs UX for single-step scheduling.

Multi-step scheduling defers output pythonization until after N gpu-side steps have completed.

For multi-step scheduling, currently the following pythonization steps are skipped entirely rather than being deferred:
* The output of `_sample_with_torch()` in sample.py is never pythonized
* `Sampler.forward()` in sample.py never pythonizes `logprobs` (this depends on the pythonized output of `_sample_with_torch()`)
* `_pythonize_sampler_output()` in multi_step_model_runner.py never utilizes the logprobs computed by the sampler, passing a dummy value back for logprobs instead.

Note that during `profile_run()`, the statements above do not apply because pythonization is performed immediately and not deferred. It is only during actual prefill and decode steps where the above difficulties with deferred pythonization apply.

This PR adds logprobs to the multi-step UX by:
* (During `profile_run()`): having `_pythonize_sampler_output()` extract the already-computed logprobs and inject them into the pythonized output
* (During prefill/decode): having `_pythonize_sampler_output()` invoke the deferred pythonization steps & inject the resultant logprobs into the pythonized output

Related to #7528 (only the logprobs support issue)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by SolitaryThinker** [comment]
fyi #7000 has been merged to main

**Comment by afeldman-nm** [comment]
FYI, the PR as it stands moves `SamplerOutput` from `sequences.py` to `sampler.py` in order to get around issues with circular imports (this also is a sensible reorganization on principle, in my opinion.) Since every model imports sampler, there are a large number of files changed - however most of these are trivial changes to the sampler import.

**Comment by afeldman-nm** [comment]
@SolitaryThinker @comaniac

File: vllm/worker/multi_step_model_runner.py
Comment (comaniac) [inline vllm/worker/multi_step_model_runner.py:449]: Please use the following docstring style in all added functions.
```suggestion
    """Perform deferred logprob Pythonization.

    1. Pythonize GPU-side sampler result tensors into CPU-side sampler result.
    2. Pythonize GPU-side logprobs tensor into CPU-side logprobs lists,
       utilizing  the Pythonized sampler result computed in step 1.
    
    These deferred computations are not required for single-step scheduling
    or the `profile_run()` phase of multi-step scheduling.

    Args:
        output: sampler output (under deferred Pythonization)
        sampling_metadata

    Returns:
        prompt_logprobs (CPU), sample_logprobs (CPU)
    """
```

File: tests/multi_step/test_correctness.py
Comment (comaniac) [inline tests/multi_step/test_correctness.py:26]: Seems like this is only used once. In this case we could just embed it.

File: tests/multi_step/test_correctness.py
Comment (comaniac) [inline tests/multi_step/test_correctness.py:162]:
```suggestion
@pytest.mark.parametrize("num_logprobs", [None, 5])
```

File: tests/multi_step/test_correctness.py
Comment (comaniac) [inline tests/multi_step/test_correctness.py:29]: We should have these utilities ready in conftest so no need to create duplicated ones.

File: vllm/model_executor/layers/sampler.py
Comment (comaniac) [inline vllm/model_executor/layers/sampler.py:19]: Move to the top as dataclass is Python build package.
```python
import warnings
from dataclasses import dataclass
from importlib.util import find_spec
```

File: vllm/model_executor/layers/sampler.py
Comment (comaniac) [inline vllm/model_executor/layers/sampler.py:300]:
```suggestion
            # Pythonize logprobs now (GPU -> CPU); do not defer.
```

**Review by comaniac — COMMENTED** [review]
Overall LGTM. Comments are mostly for coding style and refactoring.
Also can you run some benchmarks to make sure there's no performance regression with logprobs is not requested?

File: tests/multi_step/test_correctness.py
Comment (afeldman-nm) [reply tests/multi_step/test_correctness.py]: Done.

File: tests/multi_step/test_correctness.py
Comment (afeldman-nm) [reply tests/multi_step/test_correctness.py]: Done.

File: tests/multi_step/test_correctness.py
Comment (afeldman-nm) [reply tests/multi_step/test_correctness.py]: Moved them to tests/utils.py (did not see a clear benefit of using conftest vs utils.py)

File: vllm/model_executor/layers/sampler.py
Comment (afeldman-nm) [reply vllm/model_executor/layers/sampler.py]: Fixed

File: vllm/worker/multi_step_model_runner.py
Comment (afeldman-nm) [reply vllm/worker/multi_step_model_runner.py]: Thanks, applied this change where needed.

File: vllm/worker/multi_step_model_runner.py
Comment (SolitaryThinker) [inline vllm/worker/multi_step_model_runner.py:462]: I see you clear `logprobs` here, but since this method may be deferred by a few steps, outputs will still have reference to this GPU tensor as it is returned.

**Review by SolitaryThinker — COMMENTED** [review]
LGTM, let me know if you want to discuss more about clearing `logprobs` tensor in SamplerOutput

File: vllm/worker/multi_step_model_runner.py
Comment (afeldman-nm) [reply vllm/worker/multi_step_model_runner.py:301]: Implemented a fix.

File: vllm/worker/multi_step_model_runner.py
Comment (afeldman-nm) [reply vllm/worker/multi_step_model_runner.py]: Implemented a fix.

File: tests/multi_step/test_correctness.py
Comment (comaniac) [reply tests/multi_step/test_correctness.py]: Sorry I didn't mean to move these utilities. I meant we already have something like `vllm_runner` in conftest that launches an engine for you to test. Please look at existing tests to make use of these utilities so that you don't need to create a new one.

Comment (afeldman-nm) [reply tests/multi_step/test_correctness.py]: @comaniac ahh, yes I am very familiar with VllmRunner in conftest. However those existing conftest utilities do not serve the needed purpose.

When @SolitaryThinker wrote the multi-step test (test_correctness.py), he made it an end-to-end test. He starts up a separate server process, obtains a client for it, and sends completion requests to the OpenAI endpoint.

I think we do not have a pre-existing utility for this (probably why Will wrote it from scratch.) Certainly, looking through the other tests I have not found such a utility. That is why I moved Will's code to utils.py, so that it will become a utility.

VllmRunner is not an end-to-end test utility & I am pretty sure it relies on the sync API.

**Review by SolitaryThinker — APPROVED** [review]
Thanks for the work!

File: tests/multi_step/test_correctness.py
Comment (comaniac) [inline tests/multi_step/test_correctness.py:42]: nit: Use double quotes.
```suggestion
    """Test vLLM engine with multi-step scheduling in an OpenAI-protocol
```

File: tests/utils.py
Comment (comaniac) [inline tests/utils.py:481]: We have `check_logprobs_close` in `test/models/utils.py` too. Could we merge these 2 together?

**Review by comaniac — COMMENTED** [review]
LGTM. Only nits. Thanks!
Will approve and merge once we confirm there's no performance regression.

File: tests/utils.py
Comment (afeldman-nm) [reply tests/utils.py:481]: I appreciate what you are getting at with this! However, it is worth noting that `check_logprobs_close()` has a significant difference in its logic & purpose.

`check_logprobs_close()` is designed for a scenario where you are testing for failures in model design compared to a HuggingFace baseline. It actually allows for significant differences in logprob values, because there can be differences in numerics & order-of-operations between vLLM and HuggingFace which lead to significant rounding errors. It does not test for scenarios where the output tokens are correct but the logprobs are wrong (which matters, given that my PR is adding logprobs support.) `check_logprobs_close()` *assumes* that if the tokens match, the logprobs match; therefore `check_logprobs_close()` *only* examines logprobs at a given sequence position if the tokens at that position don't match. Furthermore, when examining logprobs, `check_logprobs_close()` does not directly compare logprobs to ensure they are nearly identical; it only requires that the token in sequence A is in the top-K logprobs of sequence B, and visa versa. You can see this below:

```
def check_logprobs_close(
    *,
    outputs_0_lst: Sequence[TokensTextLogprobs],
    outputs_1_lst: Sequence[TokensTextLogprobs],
    name_0: str,
    name_1: str,
    num_outputs_0_skip_tokens: int = 0,
    warn_on_mismatch: bool = True,
):
        ...

        # Loop through generated tokens.
        for idx, (output_id_0,
                  output_id_1) in enumerate(zip(output_ids_0, output_ids_1)):

            # If generated tokens don't match, then
            if output_id_0 != output_id_1:
                logprobs_elem_0 = logprobs_0[idx]
                logprobs_elem_1 = logprobs_1[idx]

                # Each predicted token must be in top N logprobs of the other
                fail_msg = (
                    f"Test{prompt_idx}:"
                    f"\nMatched tokens:\t{output_ids_0[:idx]}"
                    f"\n{name_0}:\t{output_str_0!r}\t{logprobs_elem_0}"
                    f"\n{name_1}:\t{output_str_1!r}\t{logprobs_elem_1}")

                assert logprobs_elem_0 is not None, fail_msg
                assert logprobs_elem_1 is not None, fail_msg
                assert output_id_0 in logprobs_elem_1, fail_msg
                assert output_id_1 in logprobs_elem_0, fail_msg
```

I would argue that `check_logprobs_close()` is actually a misnomer; a better name would be `tokens_match_or_tokens_are_mutually_in_top_logprobs()`

In contrast, the all_close function which I wrote strives to show that the multi-step top-K logprobs are identical to single-step top-K logprobs (i.e. matching with atol=1e-5 and rtol=1e-5) at all token offsets, with no regard for whether the tokens match. I think this functionality is too different from `check_logprobs_close()` to be reconciled.

Comment (afeldman-nm) [reply tests/utils.py:481]: So in short, unless you have strong objections, I would prefer to keep the code the way it is.

Comment (comaniac) [reply tests/utils.py:481]: Sure ofc. Maybe you could even help rename that one to better illustrate what it does...

**Review by comaniac — APPROVED** [review]

File: tests/utils.py
Comment (afeldman-nm) [reply tests/utils.py:481]: Agreed, done.

**Comment by afeldman-nm** [comment]
## Perf test:

### Server:
```
python  -m vllm.entrypoints.openai.api_server     --model JackFram/llama-160m --swap-space 16     --disable-log-requests  --use-v2-block-manager   --tensor-parallel-size 2  --pipeline-parallel-size 2  --gpu-memory-utilization 0.90  --num-scheduler-steps 8 --worker-use-ray 2>&1 | tee output.log
```

### Client:
```
python benchmarks/benchmark_serving.py     --backend vllm      --model JackFram/llama-160m     --dataset-name sharegpt     --dataset-path ../sharegpt.json
```

### Testbench:

```
4x NVIDIA A4000 RTX 16GB
```

### Baseline perf benchmark result (main, 029c71de11b)

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  28.08     
Total input tokens:                      245995    
Total generated tokens:                  235480    
Request throughput (req/s):              35.62     
Input token throughput (tok/s):          8761.30   
Output token throughput (tok/s):         8386.80   
---------------Time to First Token----------------
Mean TTFT (ms):                          8902.03   
Median TTFT (ms):                        7759.66   
P99 TTFT (ms):                           20250.27  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          27.63     
Median TPOT (ms):                        25.76     
P99 TPOT (ms):                           68.58     
---------------Inter-token Latency----------------
Mean ITL (ms):                           186.44    
Median ITL (ms):                         188.87    
P99 ITL (ms):                            464.41    
==================================================
```

### PR branch result (after merging in 029c71de11b)

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  28.98     
Total input tokens:                      245995    
Total generated tokens:                  235481    
Request throughput (req/s):              34.50     
Input token throughput (tok/s):          8487.41   
Output token throughput (tok/s):         8124.66   
---------------Time to First Token----------------
Mean TTFT (ms):                          9234.82   
Median TTFT (ms):                        8197.66   
P99 TTFT (ms):                           20878.33  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          28.97     
Median TPOT (ms):                        26.52     
P99 TPOT (ms):                           81.90     
---------------Inter-token Latency----------------
Mean ITL (ms):                           192.62    
Median ITL (ms):                         194.00    
P99 ITL (ms):                            465.43    
==================================================
```

**Comment by comaniac** [comment]
Thanks for the benchmark. So it looks like we do have some performance overheads even logprobs is not requested?

**Comment by afeldman-nm** [comment]
## Perf test (8x multi-step + {async output proc., logprobs}, main (ec266536b7)):

### Server (TP=1, PP=1):
```
python  -m vllm.entrypoints.openai.api_server     --model meta-llama/Meta-Llama-3-8B --swap-space 16     --disable-log-requests  --use-v2-block-manager   --tensor-parallel-size 1  --pipeline-parallel-size 1  --gpu-memory-utilization 0.90  --num-scheduler-steps 8 {--disable-async-output-proc} {--max-logprobs 5} --port 9000  --worker-use-ray 2>&1 | tee output.log
```

### Client:
```
python benchmarks/benchmark_serving.py     --backend vllm      --model JackFram/llama-160m     --dataset-name sharegpt     --dataset-path ../sharegpt.json --port 9000
```

### Testbench:

```
2x NVIDIA H100 80GB
```

### Benchmarking result (+ async output proc, - logprobs)

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  25.30     
Total input tokens:                      215196    
Total generated tokens:                  128650    
Request throughput (req/s):              39.52     
Input token throughput (tok/s):          8504.48   
Output token throughput (tok/s):         5084.21   
---------------Time to First Token----------------
Mean TTFT (ms):                          4728.12   
Median TTFT (ms):                        3337.89   
P99 TTFT (ms):                           16418.05  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          54.80     
Median TPOT (ms):                        38.10     
P99 TPOT (ms):                           393.95    
---------------Inter-token Latency----------------
Mean ITL (ms):                           257.74    
Median ITL (ms):                         233.41    
P99 ITL (ms):                            699.12    
==================================================
```

### Benchmarking result (- async output proc, - logprobs)

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  29.40     
Total input tokens:                      215196    
Total generated tokens:                  128425    
Request throughput (req/s):              34.02     
Input token throughput (tok/s):          7320.12   
Output token throughput (tok/s):         4368.52   
---------------Time to First Token----------------
Mean TTFT (ms):                          5233.60   
Median TTFT (ms):                        3499.40   
P99 TTFT (ms):                           18781.88  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          62.68     
Median TPOT (ms):                        44.98     
P99 TPOT (ms):                           460.72    
---------------Inter-token Latency----------------
Mean ITL (ms):                           302.97    
Median ITL (ms):                         292.59    
P99 ITL (ms):                            766.55    
==================================================
```

### Benchmarking result (+ async output proc, + logprobs)

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  24.90     
Total input tokens:                      215196    
Total generated tokens:                  128530    
Request throughput (req/s):              40.16     
Input token throughput (tok/s):          8642.29   
Output token throughput (tok/s):         5161.78   
---------------Time to First Token----------------
Mean TTFT (ms):                          4608.57   
Median TTFT (ms):                        3219.77   
P99 TTFT (ms):                           16126.21  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          54.95     
Median TPOT (ms):                        37.56     
P99 TPOT (ms):                           418.74    
---------------Inter-token Latency----------------
Mean ITL (ms):                           255.12    
Median ITL (ms):                         237.80    
P99 ITL (ms):                            662.34    
==================================================
```

### Benchmarking result (- async output proc, + logprobs)

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  27.92     
Total input tokens:                      215196    
Total generated tokens:                  128633    
Request throughput (req/s):              35.81     
Input token throughput (tok/s):          7707.09   
Output token throughput (tok/s):         4606.90   
---------------Time to First Token----------------
Mean TTFT (ms):                          4934.27   
Median TTFT (ms):                        3297.67   
P99 TTFT (ms):                           17476.72  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          60.40     
Median TPOT (ms):                        41.65     
P99 TPOT (ms):                           427.19    
---------------Inter-token Latency----------------
Mean ITL (ms):                           284.79    
Median ITL (ms):                         282.74    
P99 ITL (ms):                            658.36    
==================================================
```

**Pull request merged**
